enclave_manager/avalon_enclave_manager/singleton/singleton_enclave_manager.py

```diff
@@ -82,53 +82,43 @@ def _create_signup_data(self):
     def _process_work_order_sync(self, process_wo_id):
         """
         Process the work-order of the specified work-order id
-        and return the response.
-        Used for synchronous execution.
+        and return the response. Used for synchronous execution.
+
         Parameters:
-        - kv_helper is lmdb instance to access database
-        - process_wo_id is id of the work-order that is to be processed
+            @param process_wo_id - Id of the work-order that is to be processed
         """
-        logger.info("Processing work order")
-        try:
-            # Get all workorders requests from KV storage lookup and process
-            list_of_workorders = self._kv_helper.lookup("wo-scheduled")
-            if not list_of_workorders:
-                logger.info("Received empty list of work orders from " +
-                            "wo-scheduled table")
-                return
-        except Exception as e:
-            logger.error("Problem while getting ids from wo-scheduled table")
-            return
-
-        for wo_id in list_of_workorders:
-            if wo_id == process_wo_id:
-                resp = self._process_work_order_by_id(wo_id)
-                return resp
+        logger.info("Processing work orders found in wo-worker-pending table.")
+
+        wo_id = self._kv_helper.csv_pop("wo-worker-pending", self._worker_id)
+        while wo_id is not None:
+            if process_wo_id == wo_id:
+                self._kv_helper.set("wo-worker-processing",
+                                    self._worker_id, wo_id)
+                return self._process_work_order_by_id(wo_id)
             else:
                 return None
-        # end of for loop
+            wo_id = self._kv_helper.csv_pop(
+                "wo-worker-pending", self._worker_id)
```

Comment:
Yes @manojgop csv_pop would remove from the wo-worker-pending table. But with singleton, the if check is more of a double check only as there is no contending worker to pick up the same work-order.
However, even with WPE, this behavior is same as of now as sync mode is being restricted to pools with a single WPE. When there are multiple WPE contending over the same queue, this will be an issue.

Reply:
We might have multiple singleton workers and a combination of singleton and KME/WPE. I would suggest to have an API which can directly fetch `process_wo_id` from the lmdb table for sync mode. lmdb access can be a remote call over the network  and it's costly to loop over all the work orders by making multiple network calls.

Reply:
Also for WPE, it's more efficient to fetch a work order from Lmdb which this WPE  can support . For example, WPE can have a config parameter of list of workload_ids which it can support and fetch a work order id which matches the workload id .